File: CPP/hashmap/1_Two_Sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); i++){
        int diff = target - nums[i];
        if (map.count(diff)) {
            return {map[diff], i};
        }
        map[nums[i]] = i;
    }
    return {};
}

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> result = twoSum(nums, target);
    cout << "[" << result[0] << ", " << result[1] << "]" << endl;
    return 0;
}


vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); i++){
        cout << "Map contents: ";
        for (auto& p : map) {
            cout << "(" << p.first << ", " << p.second << ") ";
        }
        cout << endl;
        int diff = target - nums[i];
        if (map.count(diff)) {
            return {map[diff], i};
        }
        map[nums[i]] = i;
    }
    return {};
}

File: Python/Hashmap/1. Two Sum.py
from typing import List

class Solution1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        PrevMap = []
        n = len(nums)
        for i in range(n):
            PrevMap.append(target - nums[i])
        for j in range(n):
            if nums[j] in PrevMap and j != PrevMap.index(nums[j]):
                return [j, PrevMap.index(nums[j])]
            
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        PrevMap = {}

        for i,n in enumerate(nums):
            Diff = target - n
            if Diff in PrevMap:
                return [PrevMap[Diff],i]
            PrevMap[n]=i
            
        
if __name__ == '__main__':
    nums = [2,7,11,15]
    target = 9
    sol = Solution()
    print(sol.twoSum(nums,target))